File: combinatorics/resource/ST/project_5/bai_toan_6/Divide-and-Conquer.py
# Lop cay voi nhan va danh sach con
class TreeNode:
    def __init__(self, label):
        self.label = label                 # Luu nhan cua nut
        self.children = []                # Khoi tao danh sach con rong

# Ham chia de tri (Divide and Conquer)
def tree_edit_distance_dnc(t1, t2):
    if t1 is None and t2 is None:
        return 0
    if t1 is None:
        return 1 + sum(tree_edit_distance_dnc(None, c) for c in t2.children)
    if t2 is None:
        return 1 + sum(tree_edit_distance_dnc(c, None) for c in t1.children)

    # Tinh chi phi doi nhan
    cost = 0 if t1.label == t2.label else 1

    # Duyet ghep cac nut con
    m, n = len(t1.children), len(t2.children)

    def dfs(i, j):
        if i == m and j == n:
            return 0
        results = []
        if i < m and j < n:
            results.append(tree_edit_distance_dnc(t1.children[i], t2.children[j]) + dfs(i+1, j+1))
        if i < m:
            results.append(tree_edit_distance_dnc(t1.children[i], None) + dfs(i+1, j))
        if j < n:
            results.append(tree_edit_distance_dnc(None, t2.children[j]) + dfs(i, j+1))
        return min(results)

    return cost + dfs(0, 0)

# Ham nhap cay tu ban phim
def input_tree():
    label = input("Nhap nhan goc: ")
    node = TreeNode(label)
    n = int(input(f"So con cua {label}: "))
    for i in range(n):
        print(f"-- Con thu {i+1} cua {label}:")
        child = input_tree()
        node.children.append(child)
    return node

# Ham chinh
if __name__ == "__main__":
    print("Nhap cay 1:")
    t1 = input_tree()
    print("Nhap cay 2:")
    t2 = input_tree()
    print("Edit distance (D&C):", tree_edit_distance_dnc(t1, t2))

File: combinatorics/resource/ST/project_5/bai_toan_6/Divide-and-Conquer.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

// Dinh nghia mot nut cay
struct TreeNode {
    string label;
    vector<TreeNode*> children;
    TreeNode(string l) : label(l) {}
};

// Ham chia de tri
int treeEditDistanceDNC(TreeNode* t1, TreeNode* t2) {
    if (!t1 && !t2) return 0;
    if (!t1) {
        int cost = 1;
        for (auto c : t2->children)
            cost += treeEditDistanceDNC(nullptr, c);
        return cost;
    }
    if (!t2) {
        int cost = 1;
        for (auto c : t1->children)
            cost += treeEditDistanceDNC(c, nullptr);
        return cost;
    }

    int cost = (t1->label == t2->label) ? 0 : 1;
    int m = t1->children.size();
    int n = t2->children.size();

    // De quy ghep con
    function<int(int,int)> dfs = [&](int i, int j) -> int {
        if (i == m && j == n) return 0;
        int res = INT_MAX;
        if (i < m && j < n)
            res = min(res, treeEditDistanceDNC(t1->children[i], t2->children[j]) + dfs(i+1, j+1));
        if (i < m)
            res = min(res, treeEditDistanceDNC(t1->children[i], nullptr) + dfs(i+1, j));
        if (j < n)
            res = min(res, treeEditDistanceDNC(nullptr, t2->children[j]) + dfs(i, j+1));
        return res;
    };

    return cost + dfs(0, 0);
}

// Ham nhap cay tu ban phim
TreeNode* inputTree() {
    string label;
    cout << "Nhap nhan goc: ";
    cin >> label;
    TreeNode* node = new TreeNode(label);
    int n;
    cout << "So con cua " << label << ": ";
    cin >> n;
    for (int i = 0; i < n; ++i) {
        cout << "-- Con thu " << i+1 << " cua " << label << ":\n";
        node->children.push_back(inputTree());
    }
    return node;
}

int main() {
    cout << "Nhap cay 1:\n";
    TreeNode* t1 = inputTree();
    cout << "Nhap cay 2:\n";
    TreeNode* t2 = inputTree();
    cout << "Edit distance (D&C): " << treeEditDistanceDNC(t1, t2) << endl;
    return 0;
}